content/blog/2019-11-07-immix/index.md

```diff
@@ -0,0 +1,210 @@
+2. **Demand Driven Overflow Allocation**. Each immix allocator is paired with an *overflow allocator* that is also a contiguous allocator, however, it always uses empty blocks. If immix cannot allocate a *medium* object (an object with size greater than a line) into a block that contains free lines, immix allocates it with the overflow allocator.
+
+### Opportunistic Evacuation
+
+The mark-region collector, 
+like a mark-sweep one, 
+is non-moving and thus subject to fragmentation. 
+Immix uses a defragmentation mechanism called *opportunistic evacuation*.
+
+Actually, before the evacuation process, 
+methods are needed to determine when to trigger defragmentation and how to select candidate blocks. 
+This part is left to be discussed in implementation details, 
+let's suppose we know the set of candidate blocks for now. 
+
+This evacuation process is: 
+when the collector encounters a live object in a candidate block, 
+it *opportunistically* evacuates the object. 
+That is, it only evacuates an object if prior evacuation has not exhausted all target space. 
+And the process to allocate this object is the same as to allocate an object created by the mutator,
+except all defragmentation candidate blocks are not considered free nor partially free. 
+Besides, the collector leaves a forwarding pointer which records the address of the new location.
+And if the collector encounters references to a forwarded object, 
+it replaces them with the new location.
+
+#### Implementation Details
+
+1. **Defragmentation Trigger.** If there are one or more recyclable (free or partially free) that the allocator "did not use" or the previous collection didn't free enough space, immix triggers defragmentation at the beginning of the current collection.
+2. **Candidate Selection**. Immix selects candidate blocks with more holes, as many as possible. It maintains an array called *available histogram* AH, where AH[i] reflects space occupied in all blocks with i holes. During selection, immix starts from biggest i, select all blocks with i holes as candidates if available space is not exhausted, and decrease i, repeat this process until available space is exhausted or i becomes 0.
+3. **Headroom**. Headroom is a small number of free blocks that immix never returns to the global allocator and *only* uses for evacuating.
+4. **Pinning.** In some situations, an application may request that an object not be moved. Immix supports this pinning feature. If the application pins an object, defragmentation never moves this object.
+5. <!--**Mixing Marking and Evacuation.** ??? I don't understand this part in the paper-->
+
+# Evaluation
+
+This paper has put a lot of effort into evaluations.
+
+The benchmarks contain SPECjvm98, DaCapo suites, and pseudojbb2000. 
```

Comment:
DaCapo suites -> the DaCapo suite (there's only one suite, right?)

Reply:
Maybe worth briefly mentioning what "pseudojbb2000" is:

> Pseduojbb2000 is a fixed workload version of SPEC jbb2000 [35].

No need for details, just say that it's based on a "real" benchmark suite, SPECjbb.